### ISSUE TITLE
Stat mechanism

### ISSUE THREAD

**Issue by Vosvek** [Vosvek/ProjectRogue#5]
The stat mechanism include the stand-alone `Stat`, `Stat Effect` (and `Spell Effects`), and a `Stat` and `Effect` manager. Remember that stats can affect other stats. So a Strength stat might govern the Health stat.

**Comment by Vosvek** [comment]
Progressing, yet still deciding which approach to take (enum vs ScriptableObject; array vs hard-coded class).

**Comment by Vosvek** [comment]
Stats will be divided into 3 main categories: `Attributes`, `Vitals`, and `Skills`.

**Attributes** represent core character strengths/weakness, and are the basis of all stats (that is, they govern all stats. For example, Health is governed by Constitution, and which grants +5 health for every DiceModifier point). Attributes include `Constitution`, `Strength`, `Agility`, `Intellect`, `Wisdom`, and `Charisma`. Attribute levels currently range from 1 to 20.

**Vitals** are the core deplete-able/replenish-able stats. Vitals include `Health`, and `Stamina`.

**Skills** more specified attributes, and represent actions ranging from attacking with `Heavy Weapons` or `Archery`, the use of `Occultism` or `Miracle` magic, and `Fishing` and `Cooking`. Ideally, skills range from 5 to 100, and on level up increase by 5 points.

Numbers might still change, but this is the general idea.

There is still room for `DamageType` damage and defences. The `StatSet` mechanism could control this, as at the moment it contains `ActiveStatEffects`. However, I might refactor that in later as necessary.

**Comment by Vosvek** [comment]
Ruled: all types will be `enum`s for simplicity. This might change in the future in favour of a data-driven approach. However, for now, it's not worth catering to that case.

Array vs hard-coded class is yet to be decided.

**Comment by Vosvek** [comment]
On reconsideration it is much easier to use `ScriptableObjects` than `enums`, as they're not only easy to change as the project evolves, but it's much easier to think of Stats in terms of `BaseValue`, `EffectTotal`, and the resulting `FinalValue`.

This means that, what were previously called 'Vital Stats', the likes of Health and Stamina instead rely on two stats: `Health`, and `MaxHealth`, and `Stamina`, and `MaxStamina`.

Additionally, the `LevelUp()` function/method, wherever it is, should be in charge of pairing stats. So, for example, if MaxHealth's base value is `20 + Strength's Base` value on level up, only the function needs to know about strength adjusting the max health. In other words, the max health stat shouldn't care -- it's not its job.

This extends to stat checks for interactions. The fishing function should pair the fishing stat with the wisdom stat as it sees fit.

In summary, after much consideration, it is much easier to think of stats as a base value and a sum of effect modifiers, and neither the stats themselves or the stat and effect container component should care or be responsible for how functions want to modify and use these stats.

An update will be pushed shortly.

**Comment by Vosvek** [comment]
Created and pushed the stat mechanism. Any issues or enhancements should be new issues.
